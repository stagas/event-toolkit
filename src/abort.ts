import type { Fn } from 'everyday-types'
import { bool, toFluent } from 'to-fluent'

export class AbortOptions {
  throw = bool
  latest = bool
  timeout?: number
}

export const abort = toFluent(
  AbortOptions,
  options =>
    <P extends any[], R extends Promise<any>>(fn: (signal: AbortSignal) => Fn<P, R>) => {
      let ctrl: AbortController
      return async function(this: any, ...args: P) {
        if (options.latest && ctrl) ctrl.abort()
        ctrl = new AbortController()
        if (options.timeout != null) {
          const timeoutError = new Error('Timed out')
          try {
            return await Promise.race([
              new Promise((_, reject) => setTimeout(reject, options.timeout, timeoutError)),
              fn(ctrl.signal).apply(this, args),
            ])
          } catch (error) {
            if (error === timeoutError) {
              ctrl.abort()
              if (options.throw)
                throw error
              else
                console.warn(error)
            } else {
              if (options.throw)
                throw error
              else
                console.warn(error)
            }
          }
        } else {
          return fn(ctrl.signal).apply(this, args)
        }
      }
    }
)
