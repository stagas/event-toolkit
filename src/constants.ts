export const MouseButton = {
  Left: 1,
  Right: 2,
  Middle: 4,
}
