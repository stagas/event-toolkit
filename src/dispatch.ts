import type { DetailOf, EventKeys, Prefix, StringOf } from 'everyday-types'
import { bool, Fluent, toFluent } from 'to-fluent'

export class DispatchOptions implements CustomEventInit {
  bubbles = bool
  cancelable = bool
  composed = bool
}

export type Dispatch<T> = T & Fluent<T, Required<DispatchOptions>>

export const dispatch = toFluent(
  DispatchOptions,
  options => (<T extends EventTarget, K extends EventKeys<T>>(
    el: T,
    nameOrEvent: StringOf<K> | Event,
    detail?: DetailOf<T, Prefix<'on', K>>,
    init?: CustomEventInit,
  ) =>
    el.dispatchEvent(
      nameOrEvent instanceof Event
        ? nameOrEvent
        : new CustomEvent(nameOrEvent, { detail, ...init, ...options })
    ))
)
