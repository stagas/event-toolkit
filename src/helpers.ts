export const chain = (first: () => Promise<any>, ...rest: (() => Promise<any>)[]) =>
  () => rest.reduce((p, n) => p.then(n), first())
