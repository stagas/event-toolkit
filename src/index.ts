export * from './abort'
export * from './constants'
export * from './dispatch'
export * from './event'
export * from './helpers'
export * from './on'
export * from './queue'
export * from './task'
