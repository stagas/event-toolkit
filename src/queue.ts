import type { Fn } from 'everyday-types'
import { shallowEqual } from 'everyday-utils'
import { bool, toFluent } from 'to-fluent'
import { taskGroup, taskRun, Task } from './task'

export interface Hooks {
  before?: () => void
  after?: () => void
}

export class QueueOptions {
  first = bool
  last = bool
  next = bool

  raf = bool
  task = bool
  time = bool

  atomic = bool
  concurrency?: number

  debounce?: number
  throttle?: number

  hooks?: Hooks
}

export const wrapQueue = (options: Partial<QueueOptions> = {}) =>
  <P extends any[], R>(queueFn: Fn<P, R>): Fn<P, R extends Promise<any> ? R : Promise<R>> => {
    const initialOptions = { ...options }
    const queued: Task[] = []
    let queueOuterFn: Fn<[Fn<any, any>], any>
    let id: any
    let last: any
    let runs = false

    if (options.hooks) {
      if (options.hooks.before || options.hooks.after) {
        const real = queueFn
        queueFn = function queueFn(this: any, ...args: P) {
          options.hooks!.before?.()
          const result = real.apply(this, args)
          options.hooks!.after?.()
          return result
        }
      }
    }

    if (options.raf) queueOuterFn = requestAnimationFrame
    else if (options.task) queueOuterFn = queueMicrotask
    else if (options.time) queueOuterFn = setTimeout
    else if (options.throttle != null) {
      queueOuterFn = fn => {
        runs = true
        setTimeout(() => {
          if (!queued.length) runs = false
          fn()
        }, options.throttle)
      }
    } else if (options.debounce != null) {
      queueOuterFn = fn => {
        clearTimeout(id)
        id = setTimeout(fn, options.debounce)
      }
    } else if (options.atomic)
      queueOuterFn = fn => fn()
    else {
      // No queue function provided, return identity.
      // This is used when extending this in `event`.
      Object.assign(queueFn, {
        fn: queueFn,
        update: () => queueFn
      })
      return queueFn as any
    }

    if (options.first == null && options.last == null) {
      if (options.throttle != null) {
        options.first = true
        options.last = true
      }
      else if (options.debounce != null)
        options.last = true
      else
        options.next = true
    }

    let runningTasks = 0

    function queueNext() {
      let task: Task

      if (queued.length) {
        if (options.atomic) {
          task = queued.shift()!
          taskRun(task)
            .catch((_error: Error) => {
              //!warn _error
            })
            .finally(() => {
              if (options.concurrency) {
                runningTasks--
              }
              queueOuterFn(queueNext)
            })
          if (options.concurrency) {
            runningTasks++
            if (runningTasks < options.concurrency) {
              queueOuterFn(queueNext)
            }
          }
          return
        }
        if (options.last) {
          if (options.next) {
            const left = queued.splice(0, queued.length - 1)
            task = left.pop() ?? queued.pop()!
            taskGroup(task, left)
            last = taskRun(task)
            if (queued.length || options.throttle) {
              queueOuterFn(queueNext)
            }
            return
          } else {
            task = queued.pop()!
            taskGroup(task, queued.splice(0))
            last = taskRun(task)
            if (options.throttle) {
              queueOuterFn(queueNext)
              return
            }
          }
        } else if (options.next) {
          task = queued.shift()!
          taskGroup(task, queued.splice(0, queued.length - 1))
          queueOuterFn(queueNext)
          last = taskRun(task)
          return
        } else {
          task = Task()
          taskGroup(task, queued.splice(0))
          task.resolve(last)
        }
      }
      runs = false
    }

    function queueWrap(this: any, ...args: P) {
      //!? 'wrap called'
      const task = Task(queueFn, this, args)

      if (!runs && options.first) {
        runs = true
        if (!queued.length) {
          last = taskRun(task)
          queueOuterFn(queueNext)
          return task.promise
        }
      }

      queued.push(task)

      if (!runs || options.debounce) {
        runs = true
        queueOuterFn(queueNext)
      }

      return task.promise
    }

    queueWrap.fn = queueFn
    queueWrap.options = initialOptions

    queueWrap.update = (newFn: Fn<P, R>, newOptions: QueueOptions) => {
      //!? 'updating fn'
      if (!shallowEqual(initialOptions, newOptions)) {
        //!? 'new options', initialOptions, newOptions
        return newFn
      }
      queueFn = newFn
      //!? 'updated and returned previous wrapped'
      return queueWrap
    }

    return queueWrap as any
  }

/**
 * Decorate function `fn` with a queue function.
 *
 * The decorated function will returns a `Promise` that resolves with its result.
 * All calls will be resolved with latest result at any given quantum.
 *
 * Options:
 * - `first` => Run only first, then debounce.
 * - `last` => Run last (default behavior when using `debounce`).
 * - `next` => Run final subsequent call on next quantum (default behavior when nothing is set and not a `debounce`).
 *
 * - `raf` => Queue with `requestAnimationFrame`.
 * - `task` => Queue with `queueMicrotask`.
 * - `time` => Queue with `setTimeout`.
 *
 * - `debounce(ms)` => Debounce at specified `ms`.
 * - `throttle(ms)` => Throttle at specified `ms`.
 *
 * ```ts
 * fn = (x: number) => console.log(x)
 *
 * const cb = queue.task((x: number) => count += x)
 * cb(1) // passes (after task)
 * cb(2) // discarded
 * cb(3) // discarded
 * cb(4) // passes
 *
 * const cb = queue.task.last((x: number) => count += x)
 * cb(1) // discarded
 * cb(2) // discarded
 * cb(3) // discarded
 * cb(4) // passes
 *
 * const cb = queue.task.first((x: number) => count += x)
 * cb(1) // passes (before task)
 * cb(2) // discarded
 * cb(3) // discarded
 * cb(4) // discarded
 *
 * const cb = queue.task.first.last((x: number) => count += x)
 * cb(1) // passes (before task)
 * cb(2) // discarded
 * cb(3) // discarded
 * cb(4) // passes
 *
 * const cb = queue.task.first.last.next((x: number) => count += x)
 * cb(1) // passes (before task)
 * cb(2) // discarded
 * cb(3) // passes (after task)
 * cb(4) // passes (next task)
 *
 * const cb = queue.task.last.next((x: number) => count += x)
 * cb(1) // discarded
 * cb(2) // discarded
 * cb(3) // passes (after task)
 * cb(4) // passes (next task)
 * ```
 */
export const queue = toFluent(QueueOptions, wrapQueue)
