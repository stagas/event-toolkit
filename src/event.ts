import { EventHandler, Target } from 'everyday-types'
import { bool, toFluent } from 'to-fluent'
import { QueueOptions, wrapQueue } from './queue'

export class EventOptions extends QueueOptions {
  prevent = bool
  stop = bool
  immediate = bool
}

export const wrapEvent = (options: EventOptions) =>
  <T extends Target, E extends Event>(fn: EventHandler<T, E> = () => {}) =>
    wrapQueue(options)(
      options.prevent || options.stop || options.immediate
        ? function(this: T, e: E) {
          if (options.prevent) e.preventDefault()
          if (options.stop) {
            options.immediate
              ? e.stopImmediatePropagation()
              : e.stopPropagation()
          }
          return fn.call(this, e as any)
        }
        : fn
    )

/**
 * Decorates event handler `fn`.
 *
 * Options:
 * - `prevent` => `event.preventDefault()`
 * - `stop` => `event.stopPropagation()`
 * - `stop.immediate` => `event.stopImmediatePropagation()`
 *
 * Options inherited from `queue`:
 * - `first` => Run only first, then debounce.
 * - `last` => Run last (default behavior when using `debounce`).
 * - `next` => Run final subsequent call on next quantum (default behavior when nothing is set and not a `debounce`).
 *
 * - `raf` => Queue with `requestAnimationFrame`.
 * - `task` => Queue with `queueMicrotask`.
 * - `time` => Queue with `setTimeout`.
 *
 * - `debounce(ms)` => Debounce at specified `ms`.
 * - `throttle(ms)` => Throttle at specified `ms`.
 *
 * ```ts
 * btn.onclick = event(fn)
 * btn.onclick = event.prevent(fn)
 * btn.onclick = event.prevent.stop(fn)
 * btn.onclick = event.stop.immediate(fn)
 *
 * // examples with queue
 * btn.onclick = event.stop.raf(fn)
 * btn.onclick = event.prevent.throttle(50)(fn)
 * btn.onclick = event.debounce(100)(fn)
 * ```
 */
export const event = toFluent(EventOptions, wrapEvent)
