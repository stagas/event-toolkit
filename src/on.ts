import type { EventHandler, EventKeys, EventsOf, Fn, StringOf } from 'everyday-types'
import { Getter } from 'proxy-toolkit'
import { bool, Fluent, toFluent } from 'to-fluent'
import { EventOptions, wrapEvent } from './event'

/**
 * Removes the event listener and returns a promise used for chaining.
 * @public
 */
export type Off = () => Promise<void>
export type On<T> = T & Fluent<T, Required<OnOptions>>
export type OnEvent<T, K extends EventKeys<T>> = On<Fn<[EventHandler<T, EventsOf<T>[K]>?], Off>>
export type OnGetter<T> = { [K in EventKeys<T>]: OnEvent<T, K> }

export class OnOptions extends EventOptions implements AddEventListenerOptions {
  once = bool
  passive = bool
  capture = bool
}

const onEvent = (el: EventTarget, type: string, listener: EventHandler<any, any>, options: AddEventListenerOptions) => {
  el.addEventListener(type, listener, options)
  return () => Promise.resolve(el.removeEventListener(type, listener, options))
}

const onEventFluent = (el: EventTarget, key: string) =>
  toFluent(OnOptions, options => (listener = () => {}) => onEvent(el, key, wrapEvent(options)(listener), options))

/**
 * Adds an event listener for `el` using fluent options.
 *
 * Options:
 * - `capture` => If specified, indicates that events of this type will be dispatched to the registered listener before being dispatched to any `EventTarget` beneath it in the DOM tree.
 * - `once` => If specified, indicates that the listener should be invoked at most once after being added. The listener would be automatically removed when invoked.
 * - `passive` => Indicates that the function specified by listener will never call `preventDefault()`. If a passive listener does call `preventDefault()`, the user agent will do nothing other than generate a console warning. If not specified, defaults to `not.passive` â€“ except that in browsers other than Safari and Internet Explorer, defaults to `passive` for the `wheel`, `mousewheel`, `touchstart` and `touchmove` events.
 * - `signal(ctrl.signal)` => An `AbortSignal`. The listener will be removed when the given `AbortSignal` object's `abort()` method is called. If not specified, no `AbortSignal` is associated with the listener.
 *
 * Options inherited from `event`:
 * - `prevent` => `event.preventDefault()`
 * - `stop` => `event.stopPropagation()`
 * - `stop.immediate` => `event.stopImmediatePropagation()`
 *
 * Options inherited from `queue`:
 * - `first` => Run only first, then debounce.
 * - `last` => Run last (default behavior when using `debounce`).
 * - `next` => Run final subsequent call on next quantum (default behavior when nothing is set and not a `debounce`).
 *
 * - `raf` => Queue with `requestAnimationFrame`.
 * - `task` => Queue with `queueMicrotask`.
 * - `time` => Queue with `setTimeout`.
 *
 * - `debounce(ms)` => Debounce at specified `ms`.
 * - `throttle(ms)` => Throttle at specified `ms`.
 *
 * ```ts
 * on(btn).click(e => console.log('clicked'))
 * on(btn).click.once(e => console.log('clicked once'))
 * on(btn).wheel.not.passive(e => console.log('wheel not passive'))
 *
 * const off = on(btn).pointerdown.capture(e => console.log('pointer down'))
 * off() // removes the listener
 *
 * const offPointerMove = on(window).pointermove(e => console.log('pointer move'))
 * const offPointerUp = on(window).pointerup(e => console.log('pointer up'))
 * offPointerUp(offPointerMove) // remove both listeners shortcut syntax
 *
 * const ctrl = new AbortController()
 * on(window).pointermove.signal(ctrl.signal)(e =>
 *   console.log('runs until aborted'))
 * ctrl.abort() // removes the listener by signaling abort
 *
 * on(btn).click.prevent(e => console.log('prevented default'))
 * ```
 *
 * @public
 */

export function on<T extends EventTarget>(el: T): OnGetter<T>
export function on<T extends EventTarget, K extends EventKeys<T>>(el: T, key: K): OnEvent<T, K>
export function on<T extends EventTarget, K extends EventKeys<T>>(el: T, key?: StringOf<K>) {
  return key != null
    ? onEventFluent(el, key)
    : Getter(key => onEventFluent(el, key))
}
