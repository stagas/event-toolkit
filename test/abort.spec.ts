import { abort } from '../src/abort'

describe('abort', () => {
  it('latest', async () => {
    let times = 0
    const results: any = []
    const fn = abort.latest(
      signal =>
        async (x: number) => {
          times++
          await new Promise(resolve => setTimeout(resolve, 100))
          if (signal.aborted) return
          results.push(x)
        }
    )
    fn(1)
    fn(2)
    await fn(3)
    expect(times).toBe(3)
    expect(results).toMatchSnapshot()
  })

  it('times out', async () => {
    let times = 0
    const results: any = []
    const fn = abort.latest.timeout(200).throw(
      signal =>
        async (x: number) => {
          times++
          await new Promise(resolve => setTimeout(resolve, 300))
          if (signal.aborted) return
          results.push(x)
        }
    )
    await expect(Promise.race([fn(1), fn(2), fn(3)])).rejects.toThrow('Timed out')
    expect(times).toBe(3)
    expect(results).toMatchSnapshot()
  })
})
