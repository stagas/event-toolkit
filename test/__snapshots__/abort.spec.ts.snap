// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`abort latest 1`] = `
Array [
  3,
]
`;

exports[`abort times out 1`] = `Array []`;
